feat(wallet): better lifecycle handling of blocks and transfers search commands

IvanBelyakoff opened the pull request:
- Infinite commands started only once and never restarted, stopped on context cancellation
- Finite commands are joined into AtomicGroup to stop the rest in the group in case one command fails. Otherwise other commands in the group will continue running and the failed command is not retried to restart until other commands finish.
- Fixed goroutines leakage in case of failure of some commands

TBD: error handling in findBlocksCommand

Updates [#11074](https://github.com/status-im/status-desktop/issues/11074)

IvanBelyakoff committed:
feat(wallet): separated finite and infinite commands in transfers for
proper handling of errors and commands restart.
Now:
- Infinite commands started only once and never restarted, stoped on
context.Done.
- Finite commands are joined into AtomicGroup to stop the rest in the
  group in case one command fails. Otherwise other commands in the group
  will continue running and the failed command is not retried to
  restart. Fixed goroutine leakage in case of failure of some commands

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (16)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 94469a6d | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/1/) | 2023-12-10 14:54:21 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/1//consoleText) |
| :heavy_check_mark: | 94469a6d | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/1/) | 2023-12-10 14:55:05 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231210-145110-94469a-pr4441.zip) |
| :heavy_check_mark: | 94469a6d | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/1/) | 2023-12-10 14:56:12 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231210-145110-94469a-pr4441.zip) |
| :heavy_check_mark: | 94469a6d | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/1/) | 2023-12-10 14:56:49 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231210-145110-94469a-pr4441.aar) |
| | | | | | | |
| :heavy_check_mark: | bae26406 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/2/) | 2023-12-11 13:34:36 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231211-133131-bae264-pr4441.aar) |
| :heavy_multiplication_x: | bae26406 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/2/) | 2023-12-11 13:34:56 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/2//consoleText) |
| :heavy_check_mark: | bae26406 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/2/) | 2023-12-11 13:35:43 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231211-133135-bae264-pr4441.zip) |
| :heavy_check_mark: | bae26406 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/2/) | 2023-12-11 13:36:28 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231211-133131-bae264-pr4441.zip) |
| | | | | | | |
| :heavy_multiplication_x: | c0100f8d | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/3/) | 2023-12-13 14:11:16 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/3//consoleText) |
| :heavy_check_mark: | c0100f8d | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/3/) | 2023-12-13 14:11:48 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231213-140957-c0100f-pr4441.zip) |
| :heavy_check_mark: | c0100f8d | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/3/) | 2023-12-13 14:12:14 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231213-140953-c0100f-pr4441.aar) |
| :heavy_check_mark: | c0100f8d | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/3/) | 2023-12-13 14:14:40 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231213-140957-c0100f-pr4441.zip) |
| | | | | | | |
| :heavy_multiplication_x: | b8b0f6d1 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/4/) | 2023-12-13 14:19:31 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/4//consoleText) |
| :heavy_check_mark: | b8b0f6d1 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/4/) | 2023-12-13 14:20:21 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231213-141711-b8b0f6-pr4441.zip) |
| :heavy_check_mark: | b8b0f6d1 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/4/) | 2023-12-13 14:21:54 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231213-141712-b8b0f6-pr4441.zip) |
| :heavy_check_mark: | b8b0f6d1 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/4/) | 2023-12-13 14:23:43 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231213-141710-b8b0f6-pr4441.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 9f3eba3b | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/5/) | 2023-12-13 14:42:44 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231213-144104-9f3eba-pr4441.zip) |
| :heavy_check_mark: | 9f3eba3b | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/5/) | 2023-12-13 14:45:02 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231213-144103-9f3eba-pr4441.zip) |
| :heavy_check_mark: | 9f3eba3b | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/5/) | 2023-12-13 14:46:55 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231213-144104-9f3eba-pr4441.aar) |
| :heavy_check_mark: | 9f3eba3b | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/5/) | 2023-12-13 15:14:43 | ~33 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | b42a000f | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4441/6/) | 2023-12-13 14:46:51 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-231213-144508-b42a00-pr4441.zip) |
| :heavy_check_mark: | b42a000f | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4441/6/) | 2023-12-13 14:49:09 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-231213-144509-b42a00-pr4441.zip) |
| :heavy_check_mark: | b42a000f | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4441/6/) | 2023-12-13 14:49:21 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-231213-144659-b42a00-pr4441.aar) |
| :heavy_multiplication_x: | b42a000f | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/6/) | 2023-12-13 15:24:58 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/6//consoleText) |
| :heavy_multiplication_x: | b42a000f | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/7/) | 2023-12-13 16:15:21 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/7//consoleText) |
| :heavy_check_mark: | b42a000f | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/8/) | 2023-12-13 18:24:44 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4441/8//consoleText) |

IvanBelyakoff committed:
fix(wallet): Made an interface for BlockRangesSequentialDAO to
mock it in tests.
Made a configurable timeout interval for Commander interface.
Added tests to verify loadBlocksAndTransfers command is stopped
correctly on max errors limit reached

dlipicar commented on services/wallet/transfer/commands_sequential.go:
```diff
@@ -730,17 +775,53 @@ type loadBlocksAndTransfersCommand struct {
+func (c *loadBlocksAndTransfersCommand) incStarted() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.loops++
```
matter of preference, if it's just to have a simple thread-safe counter or boolean flag I like using atomics instead, which normally have less overhead

dlipicar reviewed (approved):
In general it looks fine. I feel there's a complex mixture of commands/goroutines being spawned from several different places, there's possibly some way to organize the code differently to more easily keep track of what happens where, what's blocking and whats async. Perhaps it can be looked into later when you have the time.

IvanBelyakoff commented:
> In general it looks fine. I feel there's a complex mixture of commands/goroutines being spawned from several different places, there's possibly some way to organize the code differently to more easily keep track of what happens where, what's blocking and whats async. Perhaps it can be looked into later when you have the time.

Agree with you. The implementation and tracking of what commands and where are spawn is somewhat complex and confusing. I might think of some way to simplify that indeed. At least now I add more tests and hope that will help us to better understand the logic and introduce changes
